File: shared/src/main/java/org/dwallach/calwatch/BatteryWrapper.java
/*
 * CalWatch
 * Copyright (C) 2014 by Dan Wallach
 * Home page: http://www.cs.rice.edu/~dwallach/calwatch/
 * Licensing: http://www.cs.rice.edu/~dwallach/calwatch/licensing.html
 */
package org.dwallach.calwatch;

import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.BatteryManager;
import android.util.Log;

/**
 * Created by dwallach on 9/1/14.
 */
public class BatteryWrapper {
    private static final String TAG = "BatteryWrapper";

    private Context context;
    private boolean isCharging;
    private float batteryPct = 1.0f;

    private static BatteryWrapper singleton;


    private BatteryWrapper(Context context) {
        this.context = context;
        singleton = this;
    }

    public static void init(Context context) {
        if(singleton == null) {
            singleton = new BatteryWrapper(context);
        } else if (context != singleton.context) {
            Log.v(TAG, "Hmm, a new context");
            singleton.context = context;
        }
        singleton.fetchStatus();
    }

    public static BatteryWrapper getSingleton() {
        if(singleton == null)
            Log.e(TAG, "BatteryWrapper not initialized properly");

        return singleton;
    }

    public void fetchStatus() {
        // and now, some code for battery measurement, largely stolen from the
        // official docs.
        // http://developer.android.com/training/monitoring-device-state/battery-monitoring.html

        try {
            IntentFilter ifilter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
            Intent batteryStatus = context.registerReceiver(null, ifilter);

            // Are we charging / charged?
            int status = batteryStatus.getIntExtra(BatteryManager.EXTRA_STATUS, -1);
            isCharging = (status == BatteryManager.BATTERY_STATUS_CHARGING || status == BatteryManager.BATTERY_STATUS_FULL);

            // How are we charging?
            int chargePlug = batteryStatus.getIntExtra(BatteryManager.EXTRA_PLUGGED, -1);
//            boolean usbCharge = chargePlug == BatteryManager.BATTERY_PLUGGED_USB;
//            boolean acCharge = chargePlug == BatteryManager.BATTERY_PLUGGED_AC;

            int level = batteryStatus.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);
            int scale = batteryStatus.getIntExtra(BatteryManager.EXTRA_SCALE, -1);

            batteryPct = level / (float) scale;
        } catch (Throwable throwable) {
            // if something fails, we really don't care; whatever value as in batteryPct from
            // last time is just fine for the next time
        }
    }

    public float getBatteryPct() {
        return batteryPct;
    }

    public boolean getIsCharging() {
        return isCharging;
    }
}

File: shared/src/main/kotlin/org/dwallach/calwatch/BatteryWrapper.kt
/*
 * CalWatch
 * Copyright (C) 2014 by Dan Wallach
 * Home page: http://www.cs.rice.edu/~dwallach/calwatch/
 * Licensing: http://www.cs.rice.edu/~dwallach/calwatch/licensing.html
 */
package org.dwallach.calwatch

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.util.Log
import java.lang.ref.WeakReference

/**
 * This class wraps all of our interaction with Android's BatteryManager. In particular, we only
 * want to make infrequent request to Android for the level of the battery, but we want to support
 * very frequent queries to this wrapper class.
 */
object BatteryWrapper {
    private const val TAG = "BatteryWrapper"

    private var contextRef = WeakReference<Context>(null)

    var isCharging: Boolean = false
        private set
    var batteryPct: Float = 1.0f
        private set

    fun fetchStatus(): Unit {
        val context = contextRef.get()
        if(context == null) {
            // no context, can't do anything
            return
        }

        // and now, some code for battery measurement, largely stolen from the
        // official docs.
        // http://developer.android.com/training/monitoring-device-state/battery-monitoring.html

        try {
            val ifilter = IntentFilter(Intent.ACTION_BATTERY_CHANGED)
            context.registerReceiver(null, ifilter).apply {
                // Are we charging / charged?
                isCharging = when(getIntExtra(BatteryManager.EXTRA_STATUS, -1)) {
                    BatteryManager.BATTERY_STATUS_CHARGING, BatteryManager.BATTERY_STATUS_FULL -> true
                    else -> false
                }

                // How are we charging?
                //            val chargePlug = getIntExtra(BatteryManager.EXTRA_PLUGGED, -1)
                //            val usbCharge = chargePlug == BatteryManager.BATTERY_PLUGGED_USB;
                //            val acCharge = chargePlug == BatteryManager.BATTERY_PLUGGED_AC;

                val level = getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
                val scale = getIntExtra(BatteryManager.EXTRA_SCALE, -1)

                batteryPct = level / scale.toFloat()
            }
        } catch (throwable: Throwable) {
            // if something fails, we really don't care; whatever value as in batteryPct from
            // last time is just fine for the next time
        }
    }

    fun init(context: Context) {
        Log.i(TAG, "init")
        contextRef = WeakReference<Context>(context)
        fetchStatus()
    }
}
